lldb/include/lldb/Target/Platform.h

```diff
@@ -1003,6 +1010,21 @@ class Platform : public PluginInterface {
   FileSpec GetModuleCacheRoot();
 };
 
+class PlatformMetadata {
+public:
+  PlatformMetadata(Debugger &debugger, const ScriptedMetadata metadata);
+  ~PlatformMetadata() = default;
+
+  Debugger &GetDebugger() const { return m_debugger; }
+  const ScriptedMetadata GetScriptedMetadata() const {
+    return m_scripted_metadata;
+  }
+
+protected:
+  Debugger &m_debugger;
```

Comment:
I'm somewhat worried about how this makes a platform debugger-specific through a back door. I take it this is needed to access the script interpreter (?) What happens if this platform is later added to a different debugger object?

Reply:
What I mean is that currently, when one creates a "normal" platform instance, it is not tied to any debugger. Later, one can add it to the platform list of any debugger (even multiple ones). This scripted platform is tied to a debugger (using this field), but it can still be inserted into the platform list of any debugger -- and it's not clear to me whether that's expected to work. In principle, I guess it can work, because there's a difference between the debugger whom you are serving, and the debugger which holds your script interpreter (and the two need not be the same in theory), but is that what you want to support?

If yes, then yeah,  I think it'd be better to hold a scriptinterpreter here.

Reply:
I don't see how passing the debugger vs. the script interpreter would make a difference if the scripted platform was created with one debugger and copied to another debugger platform's list. Even if I held on the script interpreter from the first debugger, once copied to the other one, the scripted platform will still interact with the script interpreter from the first one, which is basically the same as holding to the debugger.

Am I understanding that correctly?

I'm not sure if this is a scenario we want to support, may be @jimingham would have some opinions about this.

@labath if you have some ideas how to support this scenario, I'd love to be convinced :)

Reply:
> I don't see how passing the debugger vs. the script interpreter would make a difference if the scripted platform was created with one debugger and copied to another debugger platform's list. Even if I held on the script interpreter from the first debugger, once copied to the other one, the scripted platform will still interact with the script interpreter from the first one, which is basically the same as holding to the debugger.

It's basically the same, but it makes the intent clearer, I think.

> I'm not sure if this is a scenario we want to support, may be @jimingham would have some opinions about this.

> @labath if you have some ideas how to support this scenario, I'd love to be convinced :)

I am not saying we should support this scenario (I'm also not saying we should not). What I'm saying is that this (creating a platform with one debugger, but then adding it to another one) is something that can happen, so we should figure out what to do about it. Do we want to support it or not? Do we want to prevent it from happening or not?

I don't really have an answer to those, but I think we should have one..

Reply:
If there are platforms tied to a specific debugger, we should make sure they aren't look-upable by any other Debuggers.  Maybe we need a Debugger specific list of Platforms where we put the scripted ones, and change the platform lookup so it consults the Debugger specific ones first, then the global list.  That would have the added benefit (if you do the lookup right) of being able to wrap a known platform in a scripted one to slightly modify its behavior.

Reply:
Also discussed about that offline with @JDevlieghere and he suggested may be the scripted platforms should instantiate their own script interpreter so they're never dependent on a debugger. The issue with that is that now, the processes that will be create from the scripted platform will be restricted to use that script interpreter instead of using the target's debugger script interpreter. I'm not sure yet if this is desirable or even feasible but it's another lead we could take.

Reply:
I don't like that idea, a scripted platform is really really going to want to create scripted processes.  That will be super awkward when the scripted platform and the scripted process can't share code because they are in different interpreters.  I don't want to have to explain this to people trying to use the lldb extension points.

I think a much simpler solution is that Debuggers have a list of Platforms which is where dynamically created platforms are registered, and which it consults first when resolving "platform for name".   Debuggers really should be independent of one another or things get hard to reason about.  For instance, it would be very strange to have `command script import` in debugger A make a new Platform show up in debugger B.   

So you want to keep platforms made by a debugger confined to that debugger, and having an "dynamic platforms" or whatever list in the debugger seems a pretty straightforward way of doing that.